stan/math/gpu/kernels/lower_tri_inverse_step1.hpp

```diff
@@ -0,0 +1,79 @@
+#ifndef STAN_MATH_GPU_KERNELS_LOWER_TRI_INVERSE_STEP1_HPP
+#define STAN_MATH_GPU_KERNELS_LOWER_TRI_INVERSE_STEP1_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/gpu/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+// \cond
+const char* lower_tri_inverse_step1_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Calculates the inverse with no blocking
+     *
+     * @param[in,out] A The input matrix.
+     * @param[in, out] V A temporary matrix for storing partial calculations.
+     * @param rows The number of rows for A.
+     * @note Code is a <code>const char*</code> held in
+     * <code>lower_tri_inverse_step1_kernel_code.</code>
+     *  Used in math/gpu/lower_tri_inverse.hpp.
+     *  This kernel uses the helper macros available in helpers.cl.
+     */
+    __kernel void lower_tri_inverse_step1(__global double* A,
+                                          __global double* V, int rows) {
+      int index = get_local_id(0);
+      int group = get_group_id(0);
+      int block_size = get_local_size(0);
+      int offset = group * block_size;
+      int i = index;
+      for (int j = 0; j < block_size; j++) {
+        if (i == j) {
```

Comment:
We spawn `block_size` threads here that populate this `block_size` x `block_size` identity submatrix (each thread is "responsible" for a row). I didnt want to spawn threads  `block_size` x `block_size` threads because only  `block_size` threads are used after this step.

Good catch that this should pause the execution due to the reasons we discussed. This isnt dealt with here for 2 reasons:
- modern GPUs are actually capable of dual-issues (2 instructions per warp). You dont want to abuse this, but we are fine here on most GPUs.
- this is nowhere near the bottleneck of the inverse, so we were more focused on speeding up the step2 and step3

In order to make this more efficient, we could add another kernel that would create the N identity submatrices and would spawn `N`x`block_size`x`block_size` threads. Would you be more comfortable with that? Might make this more readable also.

Reply:
Oh, I really wasn't sure which would happen. I'm still kind of unsure - it seems like if the threads are executing in lock step on the same size blocks (due to padding), then `i` and `j` should basically always be equal at the same time across all threads, right? Thus they might actually take the same branch each time and never pause? 

I was more just trying to understand the execution model better. This code is fine, though now that you mention it, if it's easy enough to separate out subprocedures / subkernels here that might be useful in other kernels, that could be cool.

Reply:
> if it's easy enough to separate out subprocedures / subkernels here that might be useful in other kernels, that could be cool

Yes! We can have subkernels and that would make some sections of the code easier to read. I got busy this week, but I want to make another PR to do exactly this. (stackoverflow example [here](https://stackoverflow.com/a/7759844))

Reply:
I will try to explain this a bit more. Let me know if its helpful. Lets say that we only have to calculate 1 inverse of size 5x5. We would spawn 5 threads in a single thread block (they will also be part of the same warp) with `index` values (which is  `i` in this case) 0,1,2,3,4
The thread with `i` = 0 will set `1.0` on the first loop pass, the thread with `i`=1 on the second pass and so on. 

Subkernels do make this more readable yeah. 

In OpenCL 2.2 a kernel will also be able to spawn new threads (this is called dynamic parallelism). That should be fun :)